logr.go

```diff
@@ -371,12 +375,49 @@ func (l Logger) IsZero() bool {
 	return l.sink == nil
 }
 
+// WithContext stores a context in the logger. If the logger also has keys set
+// that it is meant to log from a context, then the values for those keys from
+// the given context will be added to all log entries.
+func (l Logger) WithContext(ctx context.Context) Logger {
+	l.ctx = ctx
+	return l
+}
+
+// WithContextValues stores a list of context keys and the name for them. When
+// given a context through WithContext later, the logger will extract the
+// values for these keys and log them as additional key/value pairs.
+func (l Logger) WithContextValues(keys ...ContextKey) Logger {
+	l.contextKeys = &keys
```

Comment:
Shouldn't this be an append, rather than a replace (which requires a full copy of the array)?

Reply:
Semantic aside (set vs. append), even if we append we have to do a full copy. Otherwise we may end up with the situation where an `append` in one `Logger` overwrites a value stored by the `append` in another `Logger`. Here's an example where that happens: https://go.dev/play/p/Ze2zhR0U2VS

It all depends on the capacity of the slice. We don't have control over how much `append` increases that when it needs to reallocate, so this can happen.

Besides that, append to a shared slice is not thread-safe, is it?

Reply:
Oh, yes, my sentence was unclear.  append absolutely requires us to copy the array.  That's really what I mean but it was poorly phrased.

"""
Shouldn't this be an append (which requires a manual copy of the existing data), rather than a replace?
"""

Reply:
I agree, let's append the new keys. It is then consistent with `WithValues` which also adds instead of sets the new values.